Web Inspector: CSS Grid Page Overlay support for masonry layout

rcaliman-apple opened the pull request:
#### 9a3f6dd4c6a6ec871f1fe462628902fe5756cb48
<pre>
Web Inspector: CSS Grid Page Overlay support for masonry layout
<a href="https://bugs.webkit.org/show_bug.cgi?id=265381">https://bugs.webkit.org/show_bug.cgi?id=265381</a>
<a href="https://rdar.apple.com/118833618">rdar://118833618</a>

Reviewed by Patrick Angle.

There are no actual rows or columns in the masonry axis of a masonry layout.
But we can borrow the concept to draw the two lines at the start and end of the masonry axis.

There are rows or columns in the grid axis of a masonry layout, just like in a regular grid layout.
Their lines will wrap the masonry layout in the other direction.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::buildGridOverlay):

Canonical link: <a href="https://commits.webkit.org/271332@main">https://commits.webkit.org/271332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce943562998c4bcf55690537de306740d35bd652

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29259 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30489 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25506 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28454 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8589 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3988 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25273 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28226 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5349 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24022 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4606 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4792 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25028 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31178 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25560 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25465 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31046 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4801 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28858 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6331 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6724 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5246 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5281 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1548,11 +1548,27 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+    // But we can borrow the concept to draw the two lines at the start and end of the masonry axis.
+    if (renderGrid.areMasonryRows()) {
+        auto firstRowPosition = rowPositions[0];
+        auto lastRowPosition = masonryContentSize;
```
Will the first row/column position always be 0? Or should `lastRowPosition` actually equal `firstRowPosition + masonryContentSize`?

rcaliman-apple replied:
Good catch! 👍 

No, the first row (or column) position isn't always zero because it includes the border, padding and offsets for the grid container itself:
https://github.com/WebKit/WebKit/blob/f6ce8ab19c996c796d7ae5c6fa23ab6a27e64678/Source/WebCore/rendering/RenderGrid.cpp#L1492

Fixed in the latest patch.

**Screenshot before**
<img width="1392" alt="Screenshot 2023-11-29 at 21 05 55" src="https://github.com/WebKit/WebKit/assets/105809798/74a8034c-8d25-4653-af82-c27e7d5b5794">


**Screenshot after**
<img width="1392" alt="Screenshot 2023-11-29 at 21 00 55" src="https://github.com/WebKit/WebKit/assets/105809798/40332104-9087-4aa6-93b1-c00d5b3454c8">

(the gap between the last item and the bottom line is the default margin for a `<p>` from the browser styles)

rcaliman-apple committed:
Web Inspector: CSS Grid Page Overlay support for masonry layout
https://bugs.webkit.org/show_bug.cgi?id=265381
rdar://118833618

Reviewed by Patrick Angle.

There are no actual rows or columns in the masonry axis of a masonry layout.
But we can borrow the concept to draw the two lines at the start and end of the masonry axis.

There are rows or columns in the grid axis of a masonry layout, just like in a regular grid layout.
Their lines will wrap the masonry layout in the other direction.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::buildGridOverlay):

Canonical link: https://commits.webkit.org/271332@main
